stdlib/public/core/UnicodeScalarProperties.swift

```diff
@@ -692,29 +692,52 @@ extension Unicode.Scalar.Properties {
   /// all current case mappings. In the event more space is needed, it will be
   /// allocated on the heap.
   internal func _applyMapping(_ u_strTo: _U_StrToX) -> String {
-    let count = 64
-    var array = _FixedArray16<UInt64>(allZeros: ()) // 16 * UInt64 take 128 bytes, same as 64 * UInt16
-    let string: String = array.withUnsafeMutableBufferPointer { bufPtrUInt64 in
-      let bufPtrUInt16: UnsafeMutablePointer<UInt16> = UnsafeMutableRawPointer(bufPtrUInt64.baseAddress._unsafelyUnwrappedUnchecked)
-        .bindMemory(to: UInt16.self, capacity: count)
-      let len: Int = _scalar.withUTF16CodeUnits { utf16 in
+    // Allocate 16 code units on the stack.
+    var fixedArray = _FixedArray16<UInt16>(allZeros: ())
+    let count: Int = fixedArray.withUnsafeMutableBufferPointer { buf in
+      return _scalar.withUTF16CodeUnits { utf16 in
         var err = __swift_stdlib_U_ZERO_ERROR
         let correctSize = u_strTo(
-          bufPtrUInt16,
-          Int32(count),
+          buf.baseAddress._unsafelyUnwrappedUnchecked,
+          Int32(buf.count),
           utf16.baseAddress._unsafelyUnwrappedUnchecked,
           Int32(utf16.count),
           "",
           &err)
         guard err.isSuccess else {
           fatalError("Unexpected error case-converting Unicode scalar.")
         }
-        // TODO: _internalInvariant(count == correctSize, "inconsistent ICU behavior")
         return Int(correctSize)
       }
-      return String._uncheckedFromUTF16(UnsafeBufferPointer<UInt16>(start: bufPtrUInt16, count: len))
     }
-    return string
+    if _fastPath(count <= 16) {
+      fixedArray.count = count
+      return fixedArray.withUnsafeBufferPointer {
+        String._uncheckedFromUTF16($0)
+      }
+    }
+    // Allocate `count` code units on the heap.
+    let array = Array<UInt16>(unsafeUninitializedCapacity: count) {
+      buf, initializesCount in
+      _scalar.withUTF16CodeUnits { utf16 in
+        var err = __swift_stdlib_U_ZERO_ERROR
+        let correctSize = u_strTo(
+          buf.baseAddress._unsafelyUnwrappedUnchecked,
+          Int32(buf.count),
+          utf16.baseAddress._unsafelyUnwrappedUnchecked,
+          Int32(utf16.count),
+          "",
+          &err)
+        guard err.isSuccess else {
+          fatalError("Unexpected error case-converting Unicode scalar.")
+        }
+        _internalInvariant(count == correctSize, "inconsistent ICU behavior")
+        initializesCount = Int(correctSize)
```

Comment:
```suggestion
        initializedCount = Int(correctSize)
```

Is there any advantage to setting this value here versus doing so before the `guard` statement and `_internalInvariant` check?

Reply:
fixed

Reply:
Sorry; I deleted the comment I made at first because where you had the line originally was more correct than what I suggested. We cannot actually know that `initializedCount` elements are initialized unless we know that `err` is successful.

Reply:
From https://developer.apple.com/documentation/swift/array/3200717-init:

>Inside the closure, set the initializedCount parameter to the number of elements that are initialized by the closure. The memory in the range buffer[0..<initializedCount] must be initialized at the end of the closure’s execution, and the memory in the range buffer[initializedCount...] must be uninitialized. This postcondition must hold even if the initializer closure throws an error.

How we may fulfil this requirement then?

Reply:
The closure doesn't throw; when you use `fatalError` or `_internalInvariant`, the closure never returns.